## Introduction:
In this assignment I will empirically verify the concept of confidence intervals. A confidence interval is a range of values in which we are confident the true value of a parameter lies in.

Specifically, I will be using a Normal distribution with mean $\mu = 3$ and variance $\sigma^2 = 25$, i.e. $N(3,25)$.

The theoretical $(1-\alpha)$% confidence interval for a $N(\mu,\sigma^2)$ distribution for the parameter $\mu$ is $\mu \pm z_{\alpha/2}\frac{\sigma}{\sqrt{n}}$ where n is the number of observations in the sample. Furthermore the theoretical $(1-\alpha)$% confidence interval for the parameter $\sigma^2$ is $[\frac{(n-1)\sigma^2}{\chi^2_{1-\alpha/2}},\frac{(n-1)\sigma^2}{\chi^2_{\alpha/2}}]$ where $\chi^2$ has n-1 degrees of freedom.

I will generate 100 random samples of 1000 observations (n=1000) in order to see how many times the confidence bound is breached compared to 95%, 90%, and 80% confidence.

In [20]:
import numpy as np
n= 1000
s = 100
# creating an array with 100 samples with 1000 observations each of a normal distribution
# with mean 3 and standard deviation 5
np.random.seed(834920347)
samples = np.random.normal(3,5,size =(s,n))
print("shape of samples:", samples.shape)

shape of samples: (100, 1000)


In [21]:
samples

array([[ -6.58049047,   2.82135723,   9.63384165, ...,   4.04994403,
          9.10682189,   2.26962127],
       [ 19.38101175,  -1.14068869,   0.83794923, ...,   3.13569595,
         -1.44113795,   1.18646105],
       [  0.74896961,  -5.94013537,   8.24467059, ...,  -2.33065537,
          3.72962942,   1.85064374],
       ..., 
       [ -2.20456576,   2.65097802,  14.96749523, ...,   7.97051682,
          6.49084567,  -2.24634139],
       [-12.01919328,   8.14415992,   8.84928235, ...,  -3.91341474,
          1.59928907,  -7.61376978],
       [  1.46035428,   3.75244024,  13.48327691, ...,  -6.83932634,
          4.24886132,   7.3017669 ]])

In [22]:
# find mean and variance of each sample
mean = []
var = []
for i in range(0,100):
    mean += [np.mean(samples[i])]
    var += [np.var(samples[i])]

In [23]:
print("Length of list of sample means =",len(mean))
print("Length of list of sample variance =",len(var))

Length of list of sample means = 100
Length of list of sample variance = 100


In [24]:
mean[:10] #print out first 10 sample means

[2.9330918646051329,
 2.7791053524305327,
 3.1935887365326985,
 2.9898048256619751,
 3.0791926945883374,
 3.1065544376878895,
 2.6657987131678156,
 3.1220186411171267,
 2.9636698801887777,
 2.9731572186100603]

In [25]:
var[:10] #print out first 10 sample variance

[25.895819733400575,
 24.286018569754884,
 25.288113814883893,
 23.156476747767964,
 25.003938545703114,
 24.475718056664764,
 25.295380033306238,
 24.332615009604531,
 26.234963858146834,
 23.028866109789]

In [26]:
# We use this function to count the number of sample parameters that breach the CI
def count_breach_CI(param = 1, l_bound = 0, u_bound = 1):
    count = 0
    for i in range(0,len(param)):
        if param[i] < l_bound or param[i] > u_bound:
            count += 1
    return count

### 95% Confidence Interval:
First we calculate the theoretical 95% confidence interval for $\mu$ and $\sigma^2$, so $\alpha = 0.05$.
From a Normal Curve Area table we get that $z_{0.025} = 1.96$, and from a $\chi^2$ distribution table we get $\chi^2_{0.975} = 1088.49$ and $\chi^2_{0.025} = 913.3$.

In [27]:
l_bound_mean = 3 - (1.96*(5/1000**0.5))
u_bound_mean = 3 + (1.96*(5/1000**0.5))
print("Lower Bound for mean:",round(l_bound_mean,2))
print("Upper Bound for mean:",round(u_bound_mean,2))
l_bound_var = 999*25/1088.49
u_bound_var = 999*25/913.3
print("Lower Bound for variance:",round(l_bound_var,2))
print("Upper Bound for variance:",round(u_bound_var,2))

Lower Bound for mean: 2.69
Upper Bound for mean: 3.31
Lower Bound for variance: 22.94
Upper Bound for variance: 27.35


Thus, the 95% confidence interval for the mean $\mu$ of a N(3,25) with 1000 observations is [2.69 , 3.31] and for the variance $\sigma^2$ is [22.94, 27.35].

In [28]:
count95_mean = count_breach_CI(mean, l_bound_mean, u_bound_mean)
count95_var = count_breach_CI(var, l_bound_var, u_bound_var)
print("Number of sample means outside theoretical CI = ", count95_mean)
print("Number of sample variance outside theoretical CI = ", count95_var)

Number of sample means outside theoretical CI =  4
Number of sample variance outside theoretical CI =  2


#### Conclusion
4 out of the 100 sample means and lay outside the theoretical 95% confidence interval, which means 96% of the sample means were within the confidence interval. Thus, 1% more sample means were inside the confidence interval than what the 95% confidence interval predicted.

2 out of the 100 sample variances lay outside of the theoretical 95% confidence interval, which means 98% of the sample variances lay within the confidence interval. Thus, 3% more sample variances were inside the confidence interval than what the 95% confidence interval predicted.

### 90% Confidence Interval:
Next we calculate the theoretical 90% confidence interval, so $\alpha = 0.1$.
From a Normal Curve Area table we get that $z_{0.05} = 1.645$, and from a $\chi^2$ distribution table we get $\chi^2_{0.95} = 1073.64$ and $\chi^2_{0.05} = 926.63$.

In [29]:
l_bound_mean = 3 - (1.645*(5/1000**0.5))
u_bound_mean = 3 + (1.645*(5/1000**0.5))
print("Lower Bound for mean:",round(l_bound_mean,2))
print("Upper Bound for mean:",round(u_bound_mean,2))
l_bound_var = 999*25/1073.64
u_bound_var = 999*25/926.63
print("Lower Bound for variance:",round(l_bound_var,2))
print("Upper Bound for variance:",round(u_bound_var,2))

Lower Bound for mean: 2.74
Upper Bound for mean: 3.26
Lower Bound for variance: 23.26
Upper Bound for variance: 26.95


Thus, the 90% confidence interval for the mean $\mu$ of a N(3,25) with 1000 observations is [2.74 , 3.26] and for the variance $\sigma^2$ is [23.26, 26.95].

In [30]:
count90_mean = count_breach_CI(mean, l_bound_mean, u_bound_mean)
count90_var = count_breach_CI(var, l_bound_var, u_bound_var)
print("Number of sample means outside theoretical CI = ", count90_mean)
print("Number of sample variance outside theoretical CI = ", count90_var)

Number of sample means outside theoretical CI =  10
Number of sample variance outside theoretical CI =  7


#### Conclusion
10 out of the 100 sample means lay outside of the theoretical 90% confidence interval, which means 90% of the sample means were within the confidence interval. This is exactly the same as what is predicted by the 90% confidence interval.

7 out of the 100 sample variances lay outside of the theoretical 90% confidence interval, which means 93% of sample variances were within the confidence interval. Thus, 3% more sample variances were inside the confidence interval than what the 90% confidence interval predicted.

### 80% Confidence Interval:
Last we calculate the theoretical 80% confidence interval, so $\alpha = 0.2$.
From a Normal Curve Area table we get that $z_{0.1} = 1.282$, and from a $\chi^2$ distribution table we get $\chi^2_{0.90} = 1056.695$ and $\chi^2_{0.10} = 942.16$.

In [31]:
l_bound_mean = 3 - (1.282*(5/1000**0.5))
u_bound_mean = 3 + (1.282*(5/1000**0.5))
print("Lower Bound for mean:",round(l_bound_mean,2))
print("Upper Bound for mean:",round(u_bound_mean,2))
l_bound_var = 999*25/1056.695
u_bound_var = 999*25/942.16
print("Lower Bound for variance:",round(l_bound_var,2))
print("Upper Bound for variance:",round(u_bound_var,2))

Lower Bound for mean: 2.8
Upper Bound for mean: 3.2
Lower Bound for variance: 23.64
Upper Bound for variance: 26.51


Thus, the 80% confidence interval for the mean $\mu$ of a N(3,25) with 1000 observations is [2.8 , 3.2] and for the variance $\sigma^2$ is [23.64, 26.51].

In [32]:
count90_mean = count_breach_CI(mean, l_bound_mean, u_bound_mean)
count90_var = count_breach_CI(var, l_bound_var, u_bound_var)
print("Number of sample means outside theoretical CI = ", count90_mean)
print("Number of sample variance outside theoretical CI = ", count90_var)

Number of sample means outside theoretical CI =  20
Number of sample variance outside theoretical CI =  14


#### Conclusion
20 out of the 100 sample means and lay outside of the theoretical 80% confidence interval, which means 80% of the sample means were within the confidence interval. This is exactly the same as what is predicted by the 80% confidence interval.

14 out of the 100 sample variances lay outside of the theoretical 80% confidence interval, which means 86% of the sample variances were within the confidence interval. Thus, 6% more sample variances were inside the confidence interval than what the 80% confidence interval predicted.